File: src/graph/union-find/km107.ts
import { createInterface } from "readline";

const rl = createInterface({
  input: process.stdin,
  output: process.stdout,
});

let lineno = 0;

let nodeNum: number = 0;
let edgeNum: number = 0;

const edges: [start: number, end: number][] = [];
let parents: number[] = [];
let src = 0;
let dst = 0;

rl.on("line", (line) => {
  lineno++;

  if (lineno === 1) {
    [nodeNum, edgeNum] = line
      .trim()
      .split(" ")
      .map((item) => Number.parseInt(item));
    parents = Array.from({ length: nodeNum + 1 }, (_, idx) => idx);
    return;
  }

  if (lineno == edgeNum + 2) {
    [src, dst] = line
      .trim()
      .split(" ")
      .map((item) => Number.parseInt(item));

    rl.close();
    rico();
    return;
  }

  const [start, end] = line
    .trim()
    .split(" ")
    .map((item) => Number.parseInt(item));
  edges.push([start, end]);
});

function find(idx: number): number {
  if (parents[idx] === idx) {
    return idx;
  }

  const root = find(parents[idx]);
  parents[idx] = root;
  return root;
}

function union(idxA: number, idxB: number) {
  const rootA = find(idxA);
  const rootB = find(idxB);

  if (rootA === rootB) {
    return;
  }

  parents[rootA] = rootB;
}

function rico() {
  for (const edge of edges) {
    union(edge[0], edge[1]);
  }

  const rootSrc = find(src);
  const rootDst = find(dst);
  console.log(rootSrc === rootDst ? 1 : 0);
}

File: src/graph/union-find/km107.py
from typing import List, Tuple


parentIdx: List[int]
edges: List[Tuple[int, int]] = []

nodeNum, edgeNum = map(int, input().strip().split(" "))

for _ in range(edgeNum):
    start, end = map(int, input().strip().split(" "))
    edges.append((start, end))

src, dst = map(int, input().strip().split(" "))

# parentIdx
# 元素的下标, 是每个节点的序号
# 元素的值, 是对应节点的父节点的序号
parentIdx = [idx for idx in range(nodeNum + 1)]

# union 合并


def union(idxA: int, idxB: int) -> None:
    rootA = find(idxA)
    rootB = find(idxB)
    parentIdx[rootA] = rootB


# find 找根节点
def find(idx: int) -> int:
    pIdx = parentIdx[idx]

    if pIdx == idx:
        return idx

    root = find(pIdx)
    # 路径压缩
    parentIdx[idx] = root

    return root


for edge in edges:
    idxA = edge[0]
    idxB = edge[1]

    union(idxA, idxB)

rootSrc = find(src)
rootDst = find(dst)

print(1 if rootSrc == rootDst else 0)
